SDK documentation Validation automation project init framework to setup pipeline.

zedy-wj opened the pull request:
We are working on an SDK documentation validation automation project. The part related to the Python SDK documentation has been completed, so we are planning to migrate it to this repository.

Since this tool involves a relatively large amount of content, we need to ensure the migration process goes smoothly. To do this, we will create a pipeline to test the contents of the pull request.

**This PR mainly includes the entry point (`eng/pipelines/python/start.yml`) for setting up the pipeline**. Additionally, it contains some descriptor files to outline the overall structure of the project.

The origin repo of the project: https://github.com/zedy-wj/content-validation.

After this PR is merged, we will create a new PR that includes all the code from the tool to complete the full migration.

@xiangyan99, @msyyc for notification.

weshaggard commented on tools/content-validation/eng/pipelines/python/start.yml:
```diff
@@ -0,0 +1,132 @@
+name: Run switched packages content validation for python
```
Is this specific to python?

I assume this pipeline is expected to be created in our DevOps org. Can you please name the yml file closer to what it is about? start.yml isn't very descriptive perhaps something like docs-content-validation.yml.

zedy-wj replied:
Yes, it is specific to python, in our current design, each language should have its own pipeline.

Updated to `python-sdk-docs-content-validation.yml`

weshaggard commented on tools/content-validation/eng/pipelines/python/start.yml:
```diff
@@ -0,0 +1,132 @@
+      env:
+        PackageName: ${{ parameters.packages }}
+        PipelineRunLink: $(pipelineUrl)
+        GitHubPat: $(GITHUB_PAT)
```
WEe should define these as pipeline variables in this yml and then for the PAT we will hook up the variable group to the pipeline once we create the pipeline.

zedy-wj replied:
The current code is used in our test pipeline. After you create the pipeline, we will adjust the code accordingly based on the location of the PAT.

weshaggard commented on tools/content-validation/eng/pipelines/common/content-validation-for-python.yml:
```diff
@@ -0,0 +1,191 @@
+      displayName: Build all projects and libraries
+      inputs:
+        command: 'build'
+        projects: 'content-validation.sln'
```
Are you missing this code in the PR?

zedy-wj replied:
These files are not currently included in this PR plan.

zedy-wj commented:
@weshaggard - 
As you mentioned, the current PR is indeed incomplete. However, its primary purpose is to introduce an entry file `python-sdk-doc-content-validation.yml` for setting up a pipeline. We understand that this pipeline will not be runnable immediately, as the functional code will be provided in a follow-up PR.

The main goal at this stage is to establish the pipeline infrastructure so that future PRs, which will include the actual implementation, can be properly tested. We appreciate your understanding and are happy to adjust the approach if needed.

zedy-wj commented on tools/content-validation/eng/pipelines/python/start.yml:
```diff
@@ -0,0 +1,132 @@
+  jobs:
+  - ${{ each packageName in split(parameters.packages, ',') }}:
+    - ${{ if ne(parameters.packages, 'all') }}:
+      - template: /eng/pipelines/python/package/${{ packageName }}.yml
```
These files are not currently included in this PR plan.

weshaggard commented:
After chatting with @xiangyan99 you should first get the code into the repo and a ci.yml file which is a pipeline just to build and test that code. After that is all working then you can do a PR to add another pipeline which can be used to run the automation on some schedule. It does no good to add the broken pipeline into the repo currently.

zedy-wj commented:
Hi, @weshaggard - The prerequisite PR https://github.com/Azure/azure-sdk-tools/pull/10738 has been merged. In this PR, we will complete Step2: `adding the pipeline configuration`. Currently, two features of this tool are disabled:

1. **Automatic issue creation**: we need to discuss with @xiangyan99 where these issues should be created.
2. **Generating the status markdown file**: in our personal repository, we update the status markdown file in a dedicated branch. We're not sure if this approach is appropriate for this repository. @xiangyan99 Do you have any ideas?

weshaggard commented on tools/content-validation/eng/pipelines/common/content-validation-for-dotnet.yml:
```diff
@@ -0,0 +1,151 @@
+- name: GitHubRepo
+  type: string
+  default: ''
+- name: AzureDevOpsPat
```
What is this needed for?

zedy-wj replied:
We previously had a feature: when automatically creating GitHub issues, we would list only part of the errors in the issue body, and then append a link to the pipeline artifacts at the end. Through that link, users could download an Excel file to view all the issues. However, this feature required an Azure DevOps PAT, and our PAT tokens are only valid for seven days. During the testing phase, frequent updates were needed, so we disabled this feature.

Currently, we’ve adopted an alternative approach proposed by @xiangyan99: per package per rule one issue. With this, individual issue bodies are no longer too long and thus avoid errors.

@msyyc do you think the artifact link feature is still necessary? If not, we’ll remove all PAT-related parameters, since as of now, they don’t seem to be used.

zedy-wj commented on tools/content-validation/eng/pipelines/dotnet/package/azure-ai-agents-persistent.yml:
```diff
@@ -0,0 +1,31 @@
+
```
I think you're right. Our current pipeline works in the same way — one pipeline handles the validation for all packages of a given language. You can check their format via https://dev.azure.com/v-wenjyu/content-validation-automation/_build/results?buildId=54&view=results. (I've tried to invite you; you can accept it via the email to gain access.)

As for the “per pipeline per package” approach you mentioned, I guess it's because you saw many YAML files for different packages. In reality, during execution, each package is assigned to a separate job to run, and those YAML files simply configure specific parameters for them. When creating pipelines, we only need to create four pipelines — one each for Python, Java, JS, and .NET. These correspond to the `eng/pipelines/language/xxx-sdk-docs-content-validation.yml` files.

zedy-wj replied:
Yes, this will end up in the Azure SDK DevOps instance once it is ready. Currently, there is a separate pipeline for each of the four languages:

- eng/pipelines/dotnet/dotnet-sdk-docs-content-validation.yml

- eng/pipelines/dotnet/java-sdk-docs-content-validation.yml

- eng/pipelines/dotnet/javascript-sdk-docs-content-validation.yml

- eng/pipelines/dotnet/python-sdk-docs-content-validation.yml

Regarding the "one job per package" model — we are trying to switch to a model where one job validates all packages of a single language. One thing we’d like to point out is: although the previous approach had many jobs, they ran in parallel, and each pipeline typically finished within 2–3 hours. With the new design where a single job validates all packages, the total runtime of a pipeline could extend to 15–20 hours.

weshaggard replied:
What is this tool doing that it takes 15-20 hours to validate doc files? That should be looked at more closely to see if that can be improved.

One pipeline per language seems reasonable but I would not expect this pipeline to run for more than 1-2 hours so please investigate why that is taking so long.

weshaggard replied:
What sort of checks are we doing for the links? I wouldn't expect it to take a few seconds per link that is pretty slow. Are you actually invoking each link? If so, we should try to cache that data because doing all that work every time shouldn't be necessary. For example, we have a daily pipeline https://github.com/Azure/azure-sdk-tools/blob/main/eng/pipelines/githubio-linkcheck.yml#L139 that will cache all the links that we visit from a number of our sites. You might be able to reuse that cache or create a similar cache in order to avoid a lot of this cost.

zedy-wj replied:
Sorry, I may have expressed myself unclearly and caused some misunderstanding. When I mentioned the `4,600` "doc links" that need to be validated, I wasn’t referring to checking whether the links themselves are valid or accessible, but rather to validating the content within each linked page.

You might be missing a bit of context, so let me briefly explain the actual purpose of this tool. Using `Python` as an example: the tool is designed to validate all packages listed on this page: https://learn.microsoft.com/en-us/python/api/overview/azure/?view=azure-python. So, when I mentioned “doc links,” I meant that there are approximately 4,600 documentation pages across the Python SDK that require content validation.

For Python, we’ve defined `6` validation rules, which means that each documentation page (or doc link) will generate 6 test cases. Let me illustrate with one specific rule: say we are testing for the `“Missing Content”` issue in the Python SDK documentation, for each doc link, the tool will detect and extract all such issues (like the ones shown in the screenshot), and automatically create GitHub issues to track them. This constitutes a single test case.
<img width="596" height="250" alt="image" src="https://github.com/user-attachments/assets/ab77b232-c1d1-4e98-a85b-a3143abb6f03" />

Therefore, we believe that spending a few seconds per test case is still quite reasonable. And since there are around `27,600` such test cases for the Python SDK alone, that explains why the overall execution takes a relatively long time.

As for your other question: “Are you actually invoking each link?” — the answer is yes. Before running the tests, we first retrieve the list of doc links whose content needs to be validated. The caching suggestion you mentioned is a good one. In fact, we implemented a similar mechanism in a recent PR, where all the doc links are saved into a [`DataFolder`](https://github.com/Azure/azure-sdk-tools/tree/main/tools/content-validation/DataSource/DataFolder/python).

That said, we're currently unsure how frequently these links or their content might change, so we’ve been cautious about using the cached data as a full replacement for fresh data collection each time. Based on our testing, even with local caching in place, we would likely only save about 1 hour of runtime, for example, reducing total runtime from 15 hours to 14 hours. It doesn’t significantly change the overall testing duration.

weshaggard replied:
Instead of pulling all these files via web crawling wouldn't it be more efficient to validate them from the docs repo directly? https://github.com/MicrosoftDocs/azure-docs-sdk-python

cc @danieljurek @xiangyan99

weshaggard replied:
This is moving in the right direction. Can we figure out a way to dynamically find the list of packages so we don't have to hard-code and keep that list updated?

weshaggard commented on tools/content-validation/eng/pipelines/python/run-all-packages.yml:
```diff
@@ -0,0 +1,88 @@
+  displayName: |
+    A list of package names to validate.
+  default:
+    - azure-ai-contentsafety
```
While reducing the packages to just a list here is a step in the right direction I think we can still do better. Can we figure out a way to dynamically find the list of packages so that we don't have to maintain this and we don't have to worry about missing any?

zedy-wj replied:
One way we can think of at the moment is to access this [page](https://learn.microsoft.com/en-us/python/api/overview/azure/?view=azure-python-preview) through playwright and get all the packages according to the service. However, there is a problem, not all packages in our test plan need to be tested, such as management-related packages, abandoned packages, and some packages that do not need to be tested. We need to filter these out. This means that even if we automatically get a list of packages that need to be tested, we need to configure the list of packages that need to be filtered. cc @xiangyan99 Any ideas?

weshaggard replied:
What are the criteria for filtering out? Can it be checked in code without needing a hard-coded list?

zedy-wj replied:
Due to changes in the test scope (the scope of “all” now refers to the packages released last month), we’ve updated the configuration here. It now uses dynamic fetching, and the hard-coded list has been removed.

zedy-wj commented:
@weshaggard - Sorry for the delay. We’ve removed a large number of configuration files and also replaced the hard-coded list with dynamically fetching the list of packages that need to be tested each month. Please re-review it, thanks!

weshaggard commented on tools/content-validation/eng/pipelines/python/python-sdk-docs-content-validation.yml:
```diff
@@ -0,0 +1,229 @@
+          Write-Host ">> Triggering pipeline #$pipelineId in $proj with:"
+          Write-Host $body
+
+          Invoke-RestMethod `
```
You shouldn't need to trigger the pipeline again to get the dynamic matrix. You can dynamically create the matrix like you are and then it will define the set of jobs in the same pipeline run.

weshaggard commented on tools/content-validation/PendingTestingPackagesThisMonth/Program.cs:
```diff
@@ -0,0 +1,262 @@
+                var configJson = JsonSerializer.Deserialize<List<Dictionary<string, List<string>>>>(configContent);
+                var packages = configJson?.FirstOrDefault()?["packages"] ?? new List<string>();
+
+                if (packages.Contains("azure-mgmt-*"))
```
Why are you doing filtering in code as well in the filter config files? I would expect to filter in one or the other.

zedy-wj commented:
@weshaggard - We have updated according to your comments, please re-review it, thanks.

zedy-wj commented:
@weshaggard - CI all checks have passed.
